sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs

```diff
@@ -0,0 +1,196 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class AvoidExcessiveClassCoupling : ParameterLoadingDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S1200";
+        private const string MessageFormat = "Split this class into smaller and more specialized ones to reduce its " +
+            "dependencies on other classes from {0} to the maximum authorized {1} or less.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private const int ThresholdDefaultValue = 20;
+        [RuleParameter("max", PropertyType.Integer,
+            "Maximum number of classes a single class is allowed to depend upon", ThresholdDefaultValue)]
+        public int Threshold { get; set; } = ThresholdDefaultValue;
+
+        private static ISet<KnownType> typesExcludedFromCoupling = new HashSet<KnownType>
+        {
+            KnownType.System_Boolean,
+            KnownType.System_Byte,
+            KnownType.System_SByte,
+            KnownType.System_Int16,
+            KnownType.System_UInt16,
+            KnownType.System_Int32,
+            KnownType.System_UInt32,
+            KnownType.System_Int64,
+            KnownType.System_UInt64,
+            KnownType.System_IntPtr,
+            KnownType.System_UIntPtr,
+            KnownType.System_Char,
+            KnownType.System_Single,
+            KnownType.System_Double,
+            KnownType.System_String,
+            KnownType.System_Object
```

Comment:
I made a test for fields and `Task`, `Action` and `Func` do count. I will update my tests to see if I correctly handle them (should be good). Any idea of other special types? I will make a test with pointers.